pybop/problems/base_problem.py

```diff
@@ -1,209 +1,282 @@
+        for inputs in inputs_list:
+            result = self.single_call(inputs, calculate_grad=calculate_grad)
+            results.append(result)
 
-    @property
-    def output_variables(self):
-        return list(set(self.signal + self.additional_variables))
+        return results[0] if len(inputs_list) == 1 else results
 
-    def evaluate(self, inputs: Inputs, eis=False):
+    def single_call(
+        self,
+        inputs: Inputs,
+        calculate_grad: bool,
+    ) -> float | tuple[float, np.ndarray]:
+        """Evaluate the cost and (optionally) the gradient for a single set of inputs."""
+        if calculate_grad:
+            calculate_grad = self._has_sensitivities
+
+        # Check the validity of the parameters before evaluating the cost
+        if not self.parameters.verify_inputs(inputs):
+            return self._cost.failure(calculate_grad if calculate_grad else None)
+
+        self.parameters.update(values=list(inputs.values()))
+
+        if calculate_grad:
+            y, dy = self.simulateS1(self._simulator.parameters.to_dict())
+            return self._cost.compute(y, dy=dy)
+
+        y = self.simulate(self._simulator.parameters.to_dict())
+        return self._cost.compute(y, dy=None)
+
+    def batch_call(
+        self, inputs_list: list[Inputs], calculate_grad: bool
+    ) -> list[float] | list[tuple[float, np.ndarray]]:
+        """Evaluate the cost and (optionally) the gradient for a list of inputs."""
+
+        # TODO: Upgrade the cost evaluations for batch processing
+
+        if calculate_grad:
+            costs, grads = [], []
+            for inputs in inputs_list:
+                out = self.single_call(inputs, calculate_grad=True)
+                costs.append(out[0])
+                grads.append(out[1])
+            return np.asarray(costs), np.asarray(grads)
+
+        costs = []
+        for inputs in inputs_list:
+            costs.append(self.single_call(inputs, calculate_grad=False))
+        return np.atleast_1d(costs)
+
+    def simulate(
+        self, inputs: Inputs
+    ) -> (
+        dict[str, np.ndarray]
+        | tuple[dict[str, np.ndarray], dict[str, dict[str, np.ndarray]]]
+    ):
         """
-        Evaluate the model with the given parameters and return the signal.
+        Evaluate the model with the given parameters and return the target.
 
         Parameters
         ----------
         inputs : Inputs
             Parameters for evaluation of the model.
 
-        Raises
-        ------
-        NotImplementedError
-            This method must be implemented by subclasses.
+        Returns
+        -------
+        dict[str, np.ndarray[np.float64]]
+            The simulated model output y(t), or y(ω) for EIS, for the given inputs.
         """
-        raise NotImplementedError
+        return self._simulator.simulate(inputs=inputs, calculate_sensitivities=False)
 
-    def evaluateS1(self, inputs: Inputs):
+    def simulateS1(self, inputs: Inputs):
```

Comment:
why is this function needed? Seems identical to the simulate function on the Simulator (and more appropriate there)